--- a/platform/default/src/mbgl/util/thread.cpp
+++ b/platform/default/src/mbgl/util/thread.cpp
@@ -7,6 +7,9 @@
 #include <pthread.h>
 #include <sched.h>
 #include <sys/resource.h>
+#ifdef __ANDROID__
+#include <sys/prctl.h>
+#endif
 
 namespace mbgl {
 namespace platform {
@@ -14,7 +17,11 @@ namespace platform {
 std::string getCurrentThreadName() {
     char name[32] = "unknown";
 #ifdef __linux__
+#ifdef __ANDROID__
+    prctl(PR_GET_NAME, name, 0, 0, 0);
+#else
     pthread_getname_np(pthread_self(), name, sizeof(name));
+#endif
 #endif
 
     return name;
@@ -22,11 +29,15 @@ std::string getCurrentThreadName() {
 
 void setCurrentThreadName(const std::string& name) {
 #ifdef __linux__
+#ifdef __ANDROID__
+    prctl(PR_SET_NAME, name.substr(0, 15).c_str(), 0, 0, 0);
+#else
     if (name.size() > 15) { // Linux hard limit (see manpages).
         pthread_setname_np(pthread_self(), name.substr(0, 15).c_str());
     } else {
         pthread_setname_np(pthread_self(), name.c_str());
     }
+#endif
 #endif
 }
